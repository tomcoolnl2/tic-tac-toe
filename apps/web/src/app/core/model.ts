
export interface AppState {
    appScreen: AppScreen
    appModalScreen: AppModalScreen | null
    intelligenceLevel: IntelligenceLevel
    bitBoards: [number, number]
    boardState: BoardState
    currentPlayer: PlayerSymbol
    playerSymbol: PlayerSymbol
    cpuSymbol: PlayerSymbol
    gameState: GameState
    solutionCells: [number, number, number] | null
    scores: [number, number, number]
}

export enum AppScreen {
    SETTINGS = 'settings',
    GAME = 'game',
}

export enum AppModalScreen {
    GAME_OVER = 'game-over',
    RELOAD = 'reload',
}

export interface User {
    name: string;
    avatar: PlayerSymbol;
}

export enum PlayerSymbol { X, O }   

export enum IntelligenceLevel {
    BIEBER = 'bieber',
    NOVICE = 'novice',
    MASTER = 'master',
}

export enum GameState {
    PREPLAY,
    PLAYING,
    WIN,
    LOST,
    DRAW,
}

export type BoardState = Array<PlayerSymbol | null>;

export interface LoggerState {
    appState: AppState;
    dec: number[];
    hex: string[];
    bin: string[];
}