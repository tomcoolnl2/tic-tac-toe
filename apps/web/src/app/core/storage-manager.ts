import { AppState, User } from './model';

/**
 * The StateStorage class is responsible for managing application state and user data storage.
 */
export class StateStorage {

    /** The key used to identify the data in storage. */
    public static readonly key = 'ttt';
    
    /** The underlying storage implementation (e.g., sessionStorage). */
    public storage: Storage;

    /**
     * Creates an instance of the StateStorage class.
     * @param storage - The storage implementation to use (e.g., sessionStorage).
     */
    constructor(storage?: Storage) {
        this.storage = storage || sessionStorage;
    }

    /**
     * Generates a namespaced key for storage.
     * @param ns - The namespace identifier.
     * @returns The namespaced key.
     */
    private namespace(ns: string): string {
        return `${StateStorage.key}:${ns}`;
    }

    /**
     * Retrieves the application state from storage.
     * @returns The application state.
     */
    public get state(): AppState {
        return this.fromStorage<AppState>('state');
    }

    /**
     * Stores the application state in storage.
     * @param state - The application state to store.
     */
    public set state(state: AppState) {
        this.toStorage<AppState>('state', state);
    }

    /**
     * Retrieves user data from storage.
     * @returns User data.
     */
    public get user(): User {
        return this.fromStorage<User>('user');
    }

    /**
     * Stores user data in storage.
     * @param state - The user data to store.
     */
    public set user(state: User) {
        this.toStorage<User>('user', state);
    }

    /**
     * Retrieves data from storage based on a namespace.
     * @param ns - The namespace identifier.
     * @returns The data retrieved from storage.
     */
    private fromStorage<T>(ns: string): T {
        return JSON.parse(
            this.storage.getItem(this.namespace(ns)) ?? null
        ) as T;
    }

    /**
     * Stores data in storage based on a namespace.
     * @param ns - The namespace identifier.
     * @param state - The data to store.
     */
    private toStorage<T>(ns: string, state: T): void {
        this.storage.setItem(this.namespace(ns), JSON.stringify(state));
    }

    /**
     * Clears the stored state and user data.
     */
    public clearState(): void {
        this.storage.removeItem(StateStorage.key);
    }
}
