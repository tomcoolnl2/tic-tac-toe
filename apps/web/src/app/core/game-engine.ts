import { AppModalScreen, AppState, GameState, PlayerSymbol } from './model';
import { AppStore } from './store-manager';

/**
 * The GameEngine class is responsible for managing the game logic.
 */
export class GameEngine {

    // Constants representing winning combinations and a draw mask
    public static solutionMasks = [0x1c0, 0x038, 0x007, 0x124, 0x092, 0x049, 0x111, 0x054];
    private readonly drawMask = 0x1ff; // 511


    public takeFirstTurn(state: AppState): AppState {
        const nullIndex = AppStore.getRandomNullIndex(state.boardState);
        state.boardState[nullIndex] = state.cpuSymbol;
        this.update(state, nullIndex)
        return state;
    }
    /**
     * Gets an array of winning cell indices based on a winning mask.
     * @param winningMask - The winning mask representing a winning combination.
     * @returns An array of winning cell indices.
     */
    public getWinningCells(winningMask: number): number[] {
        return Array.from({ length: 9 }, (_, i) => i).filter(i => Boolean(winningMask & (1 << i)));
    }

    /**
     * Determines the winner of the game based on the current state.
     * @param state - The current game state.
     * @returns The player symbol of the winner or null if there is no winner yet.
     */
    public determineWinner(state: AppState): PlayerSymbol | null {
        const { bitBoards, currentPlayer } = state;
        return GameEngine.solutionMasks.find(mask => (bitBoards[currentPlayer] & mask) === mask) ?? null;
    }

    /**
     * Determines if the game has ended in a draw.
     * @param state - The current game state.
     * @returns True if the game is a draw, false otherwise.
     */
    public determineDraw(state: AppState): boolean {
        const [score1, score2] = state.bitBoards;
        return (score1 | score2) === this.drawMask;
    }

    /**
     * Updates the game state based on the player's move.
     * @param state - The current game state.
     * @param cellIndex - The index of the selected cell.
     * @returns The updated game state.
     */
    public update(state: AppState, cellIndex: number): AppState {
        const mask = 1 << cellIndex;
        const [scoreX, scoreO] = state.bitBoards;

        if (!((scoreX | scoreO) & mask)) {
            state.bitBoards[state.currentPlayer] |= mask;
            state.boardState[cellIndex] = state.currentPlayer;

            const winningMask = this.determineWinner(state);
            if (winningMask !== null) {
                if (state.currentPlayer === state.playerSymbol) {
                    state.gameState = GameState.WIN;
                    state.scores[0] += 1;
                } else {
                    state.gameState = GameState.LOST;
                    state.scores[2] += 1;
                }
                state.appModalScreen = AppModalScreen.GAME_OVER;
                state.solutionCells = this.getWinningCells(winningMask) as AppState['solutionCells'];
            } else if (this.determineDraw(state)) {
                state.gameState = GameState.DRAW;
                state.scores[1] += 1;
                state.appModalScreen = AppModalScreen.GAME_OVER;
            } else {
                state.currentPlayer ^= 1 as PlayerSymbol;
            }
        }

        return state;
    }
}
