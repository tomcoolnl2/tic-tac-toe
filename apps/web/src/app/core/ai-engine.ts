import { delay } from '../utils';
import { GameEngine } from './game-engine';
import { AppState, IntelligenceLevel, PlayerSymbol } from './model';
import { AppStore } from './store-manager';

/**
 * The AIEngine class is responsible for managing AI player behavior in the game.
 */
export class AIEngine {

    /**
     * Checks if the current player can win with their next move.
     * @param {AppState} appState - The current state of the game.
     * @returns {number} The cellIndex for winning on the next move, otherwise -1.
     */
    private isWinningNextMove(appState: AppState, player: PlayerSymbol): number {
        const { bitBoards, boardState } = appState;
        const playerBitboard = bitBoards[player];
        const nullIndices = AppStore.getNullIndices(boardState);
    
        for (const cellIndex of nullIndices) {
            // Clone the bitboard and set the bit for the current empty cell
            const potentialBitboard = playerBitboard | (1 << cellIndex);
            // Check against every winning mask
            for (const winningMask of GameEngine.solutionMasks) {
                if ((potentialBitboard & winningMask) === winningMask) {
                    return cellIndex;
                }
            }
        }
    
        return -1;
    }
    

    /**
     * AI strategy for the Intelligence Level "BIEBER."
     * @param appState - The current application state.
     * @returns A number representing the selected index.
     */
    private [IntelligenceLevel.BIEBER](appState: AppState): number {
        return AppStore.getRandomNullIndex(appState.boardState) ?? 0;
    }

    // private calculateBestMove(appState: AppState): number {
    //     let bestMoveIndex = -1;
    //     let closestDistance = Infinity;
    
    //     const nullIndices = AppStore.getNullIndices(appState.boardState)

    //     for (const cellIndex of nullIndices) {
    //         // Clone the bitboard and set the bit for the current empty cell
    //         const cpuBitboard = appState.bitBoards[appState.cpuSymbol];
    //         const potentialBitboard = cpuBitboard | (1 << cellIndex);
    
    //         // Calculate the closest distance to any winning mask
    //         let minDistance = Infinity;
    //         for (const winningMask of solutionMasks) {
    //             const distance = calculateDistance(potentialBitboard, winningMask); // Custom function to calculate distance
    //             if (distance < minDistance) {
    //                 minDistance = distance;
    //             }
    //         }
    
    //         // Update the best move if the current move is closer to a winning mask
    //         if (minDistance < closestDistance) {
    //             closestDistance = minDistance;
    //             bestMoveIndex = cellIndex;
    //         }
    //     }
    
    //     return bestMoveIndex;
    // }
    
    /**
     * AI strategy for the Intelligence Level "NOVICE."
     * NOVICE will lookout for easy winning combinations 
     * and prevent the Player from winning if it is able on the next turn.
     * @param appState - The current application state.
     * @returns A Promise that resolves to void.
     */
    private [IntelligenceLevel.NOVICE](appState: AppState): number {
        let cellIndex = this.isWinningNextMove(appState, appState.playerSymbol);
        if (!~cellIndex) {
            cellIndex = AppStore.getRandomNullIndex(appState.boardState);
        }
        return cellIndex;
    }
    
    /**
     * AI strategy for the Intelligence Level "MASTER."
     * MASTER will actively block winning options for the Player
     * and actively play for winning its self
     * @param appState - The current application state.
     */
    private [IntelligenceLevel.MASTER](appState: AppState): number {
        
        let cellIndex: number;
        // check if AI can win
        cellIndex = this.isWinningNextMove(appState, appState.cpuSymbol);
        // if not, prevent human from winning
        if (!~cellIndex) {
            cellIndex = this.isWinningNextMove(appState, appState.playerSymbol);
        }
        // Choose a random next option
        // TODO Find the next best option, threaten to win
        if (!~cellIndex) {
            // cellIndex = this.calculateBestMove(appState);
            cellIndex = AppStore.getRandomNullIndex(appState.boardState);
        }
        
        return cellIndex;
    }

    /**
     * Updates the AI player's move based on the current application state.
     * @param appState - The current application state.
     * @returns A Promise that resolves to the selected index.
     */
    public async update(appState: AppState): Promise<number> {
        await delay(500);
        const index = this[appState.intelligenceLevel](appState) as number;
        return index;
    }
}
