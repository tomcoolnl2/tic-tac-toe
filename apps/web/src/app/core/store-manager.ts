import * as Rx from 'rxjs';
import { AppState, LoggerState, PlayerSymbol, GameState, User, AppScreen, IntelligenceLevel } from './model';
import { StateStorage } from './storage-manager';
import { GameEngine } from './game-engine';
import { AIEngine } from './ai-engine';

/**
 * Represents the main application state manager.
 */
class Store {
    private static _instance: Store = null;    

    private aiEngine: AIEngine = null;
    public gameEngine: GameEngine = null;
    public storage: StateStorage = null;

    /**
     * Represents the state of the application.
     */
    public state$: Rx.BehaviorSubject<AppState> = null;

    /**
     * Represents the user-specific state.
     */
    public user$: Rx.BehaviorSubject<User> = null;

    private constructor() {
        this.storage = new StateStorage();
        this.aiEngine = new AIEngine();
        this.gameEngine = new GameEngine();

        // Initialize state and user BehaviorSubjects with default or stored values
        this.state$ = new Rx.BehaviorSubject(this.storage.state ?? this.initialState);
        this.user$ = new Rx.BehaviorSubject(this.storage.user ?? this.initialUserState);
    }

    /**
     * Gets the singleton instance of the Store.
     */
    public static get instance(): Store {
        if (!Store._instance) {
            Store._instance = new Store();
        }
        return Store._instance;
    }

    /**
     * Gets the initial user state.
     */
    private get initialUserState(): User {
        return {
            name: '',
            avatar: PlayerSymbol.X,
        };
    }

    /**
     * Gets the initial application state.
     */
    public get initialState(): AppState {
        return {
            appScreen: AppScreen.SETTINGS,
            appModalScreen: null,
            intelligenceLevel: IntelligenceLevel.BIEBER,
            bitBoards: [0x0, 0x0],
            boardState: Array.from({ length: 9 }, () => null),
            currentPlayer: 0,
            playerSymbol: PlayerSymbol.X,
            cpuSymbol: PlayerSymbol.O,
            gameState: GameState.PREPLAY,
            solutionCells: null,
            scores: [0, 0, 0]
        };
    }

    /**
     * Generates the next game state based on the previous application state.
     * @param prevAppState - The previous application state.
     * @returns The next application state.
     */
    public getNextGameState(prevAppState: AppState) {
        return {
            ...this.initialState,
            playerSymbol: prevAppState.playerSymbol,
            cpuSymbol: prevAppState.cpuSymbol,
            intelligenceLevel: prevAppState.intelligenceLevel,
            scores: prevAppState.scores,
            gameState: GameState.PLAYING,
            appScreen: AppScreen.GAME,
        }
    }

    /**
     * Updates the application state and stores it.
     * @param newState - The new application state.
     */
    public nextState(newState: AppState): void {
        this.state$.next(newState);
        this.storage.state = newState;
    }

    /**
     * Resets the application state to its initial state.
     */
    public resetState() {
        this.nextState(this.initialState);
    }

    /**
     * Subscribes a React component to changes in the application state.
     * @param setState - The React state setter function.
     */
    public subscribe(setState: React.Dispatch<React.SetStateAction<AppState>>): void {
        this.state$.subscribe(setState);
    }

    /**
     * Updates the application state based on the selected cell index.
     * @param cellIndex - The index of the selected cell.
     */
    public update = async (cellIndex: number): Promise<void> => {
        await Rx.lastValueFrom(
            Rx.from(this.state$)
                .pipe(
                    Rx.take(1), // Take the latest value from the BehaviorSubject
                    Rx.mergeMap(async (state) => {
                        // Call game engine method passing current state and cellIndex
                        let newState = this.gameEngine.update({ ...state }, cellIndex);
                        // Update state using nextState method and await the completion
                        this.nextState(newState);

                        if (newState.gameState === GameState.PLAYING) {
                            // Initiate AI's turn with a delay
                            const nullIndex = await this.aiEngine.update(newState);
                            newState = this.gameEngine.update({ ...newState }, nullIndex);
                            this.nextState(newState);
                        }
                    })
                )
        );
    }
    
    /**
     * Gets the indices of null (empty) cells in the game board state.
     * @param {AppState['boardState']} boardState - The current state of the game board.
     * @returns {number[]} An array of indices representing null (empty) cells.
     */
    public getNullIndices(boardState: AppState['boardState']): number[] {
        return boardState.reduce((indices, value, index) => (value === null ? [...indices, index] : indices), []);
    }

    /**
     * Generates a random null index from the given board state.
     * @param boardState - The game board state.
     * @returns A random null index or null if none is available.
     */
    public getRandomNullIndex(boardState: AppState['boardState']): number | null {

        const nullIndices = this.getNullIndices(boardState);
        
        if (nullIndices.length === 0) {
            return null;
        }

        const randomIndex = nullIndices[Math.floor(Math.random() * nullIndices.length)];
        return randomIndex;
    }

    /**
     * Converts a number to a hexadecimal string.
     * @param n - The number to convert.
     * @returns The hexadecimal representation of the number.
     */
    public static toHexString = (n: number): string => `0x${n.toString(16)}`;

    /**
     * Converts a number to a binary string.
     * @param n - The number to convert.
     * @returns The binary representation of the number.
     */
    public static toBinString = (n: number): string => `0b${(n >>> 0).toString(2)}`;

    /**
     * Logs the application state and its representations.
     * @returns The logger state object containing various representations of the state.
     */
    public log(): LoggerState {
        const bitBoards = this.state$.value.bitBoards;
        return {
            appState: this.state$.value,
            dec: bitBoards,
            hex: bitBoards.map(Store.toHexString),
            bin: bitBoards.map(Store.toBinString),
        };
    }
}

const AppStore = Store.instance;
export { AppStore };