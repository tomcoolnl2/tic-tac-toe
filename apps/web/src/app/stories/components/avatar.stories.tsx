import type { StoryFn, Meta } from '@storybook/react';
import { PlayerSymbol } from '../../core/model';
import { Avatar, Props } from '../../components/avatar/avatar';


const Story: Meta<typeof Avatar> = {
    title: 'Components/Avatar',
    component: Avatar,
};
export default Story;

const Template: StoryFn<Props> = (args) => <Avatar {...args} />;

export const Default = Template.bind({});
Default.args = {
    type: PlayerSymbol.X,
    size: 'm',
};

export const LargeAvatar = Template.bind({});
LargeAvatar.args = {
    type: PlayerSymbol.O,
    size: 'xl',
};

export const DarkVariant = Template.bind({});
DarkVariant.args = {
    type: PlayerSymbol.X,
    size: 's',
    variant: 'dark',
};

export const LightVariant = Template.bind({});
LightVariant.args = {
    type: PlayerSymbol.O,
    size: 'l',
    variant: 'light',
};