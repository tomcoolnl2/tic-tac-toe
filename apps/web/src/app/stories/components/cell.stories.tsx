import { StoryFn, type Meta } from '@storybook/react';
import { PlayerSymbol } from '../../core/model';
import { Cell, Props } from '../../components/cell/cell';


const Story: Meta<typeof Cell> = {
    component: Cell,
    title: 'Components/Cell',
};
export default Story;


const Template: StoryFn<Props> = (args) => <Cell {...args} />;

export const CellX = Template.bind({});
CellX.args = {
    type: PlayerSymbol.X,
    index: 1,
    solutionCells: [0, 0, 0],
};

export const CellO = Template.bind({});
CellO.args = {
    type: PlayerSymbol.O,
    index: 2,
    solutionCells: [0, 0, 0],
};

export const InvertedCellX = Template.bind({});
InvertedCellX.args = {
    type: PlayerSymbol.X,
    index: 2,
    solutionCells: [0, 1, 2],
};

export const InvertedCellO = Template.bind({});
InvertedCellO.args = {
    type: PlayerSymbol.O,
    index: 2,
    solutionCells: [0, 1, 2],
};