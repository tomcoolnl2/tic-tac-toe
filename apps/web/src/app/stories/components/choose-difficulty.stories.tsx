import { StoryFn, type Meta } from '@storybook/react';
import { IntelligenceLevel } from '../../core/model';
import { ChooseDifficulty, Props } from '../../components/choose-difficulty/choose-difficulty';
import { withContainer, type ContainerProps } from '../helpers';


const EnchantedChooseDifficulty: React.FC<Props & ContainerProps> = withContainer(ChooseDifficulty, 320);
const Story: Meta<typeof ChooseDifficulty> = {
    component: ChooseDifficulty,
    title: 'Components/ChooseDifficulty',
};
export default Story;


const Template: StoryFn<Props> = (args) => <EnchantedChooseDifficulty {...args} />;

export const ChooseDifficultyEasy = Template.bind({});
ChooseDifficultyEasy.args = {
    selectedDifficultySetting: IntelligenceLevel.BIEBER
};

export const ChooseDifficultyNovice = Template.bind({});
ChooseDifficultyNovice.args = {
    selectedDifficultySetting: IntelligenceLevel.NOVICE
};

export const ChooseDifficultyMaster = Template.bind({});
ChooseDifficultyMaster.args = {
    selectedDifficultySetting: IntelligenceLevel.MASTER
};