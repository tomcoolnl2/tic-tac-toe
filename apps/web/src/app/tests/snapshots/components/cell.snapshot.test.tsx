
import { render } from '@testing-library/react'
import { PlayerSymbol } from '../../../core/model'
import { Cell } from '../../../components'

describe('Cell component snapshot test', () => {
    it('should match the snapshot', () => {
        const { asFragment } = render(<Cell type={PlayerSymbol.X} index={0} solutionCells={[0, 0, 0]} disabled={false} />);
        expect(asFragment()).toMatchSnapshot();
    });
});