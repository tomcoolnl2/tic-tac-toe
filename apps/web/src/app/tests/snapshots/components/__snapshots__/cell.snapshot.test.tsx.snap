// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell component snapshot test should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="cell invert invert-x"
    data-index="0"
    data-testid="cell-0"
    disabled=""
    value="0"
  >
    <span
      class="avatar avatar-x avatar-size-xl"
    />
  </button>
</DocumentFragment>
`;
